tests/unit/generator_test.cc

```diff
@@ -0,0 +1,236 @@
+    co_yield "[";
+    std::string s;
+    for (auto sv : strings) {
+        s = sv;
+        s.push_back(',');
+        co_yield s;
+    }
+    co_yield "]";
+}
+
+SEASTAR_TEST_CASE(test_async_generator_value_reference) {
+    using namespace std::literals;
+    std::vector<std::string_view> expected_quoted = {"["sv, "foo,"sv, "bar,"sv, "]"sv};
+    auto actual_quoted = generate_value_and_ref({"foo"sv, "bar"sv});
+    auto actual = co_await actual_quoted.begin();
+    for (auto expected : expected_quoted) {
+        BOOST_REQUIRE_EQUAL(*actual, expected);
+        co_await ++actual;
+    }
+}
+
+coroutine::experimental::async_generator<std::string>
+generate_yield_rvalue_reference(const std::vector<std::string> strings) {
+    for (auto& s: strings) {
+        co_yield s;
+    }
+}
+
+SEASTAR_TEST_CASE(test_async_generator_rvalue_reference) {
+    std::vector<std::string> expected_strings = {"hello", "world"};
+    auto actual_strings = generate_yield_rvalue_reference(expected_strings);
+    auto actual = co_await actual_strings.begin();
+    for (auto expected : expected_strings) {
+        BOOST_REQUIRE_EQUAL(*actual, expected);
+        co_await ++actual;
+    }
+}
+
+SEASTAR_TEST_CASE(test_async_generator_move_ctor) {
+    constexpr int count = 4;
+    auto actual_fibs = async_fibonacci_sequence(count, do_suspend::no);
+    return verify_fib_drained(std::move(actual_fibs), count);
+}
+
+SEASTAR_TEST_CASE(test_async_generator_swap) {
+    int count_a = 4;
+    int count_b = 42;
+    auto fibs_a = async_fibonacci_sequence(count_a, do_suspend::no);
+    auto fibs_b = async_fibonacci_sequence(count_b, do_suspend::no);
+    std::swap(fibs_a, fibs_b);
+    std::swap(count_a, count_b);
+    co_await verify_fib_drained(std::move(fibs_a), count_a);
+    co_await verify_fib_drained(std::move(fibs_b), count_b);
+}
+
+struct counter_t {
+    int n;
+    int* count;
+    counter_t(counter_t&& other) noexcept
+        : n{std::exchange(other.n, -1)},
+          count{std::exchange(other.count, nullptr)}
+    {}
+    counter_t(int n, int* count) noexcept
+        : n{n}, count{count} {
+        ++(*count);
+    }
+    ~counter_t() noexcept {
+        if (count) {
+            --(*count);
+        }
+    }
+};
+
+std::ostream& operator<<(std::ostream& os, const counter_t& c) {
+    return os << c.n;
+}
+
+coroutine::experimental::async_generator<counter_t>
+fiddle(int n, int* total) {
+    int i = 0;
+    while (true) {
+        if (i++ == n) {
+            throw std::invalid_argument("Eureka from generator!");
+        }
+        co_yield counter_t{i, total};
+    }
+}
+
+SEASTAR_TEST_CASE(test_async_generator_throws_from_generator) {
+    int total = 0;
+    auto count_to = [total=&total](unsigned n) -> seastar::future<> {
+        auto count = fiddle(n, total);
+        auto it = co_await count.begin();
+        for (unsigned i = 0; i < 2 * n; i++) {
+            co_await ++it;
+        }
+    };
```

Comment:
Maybe I should pause the internals review and look at the interface. Why an iterator range interface? It can't be be used in a range for loop. How do you determine the generator is exhausted? Here you know it generates 2*n, but in general, you don't.

I think we should go back to std::optional\<T>

Reply:
With such an interface, you could implement both future<> operator(), and operator co_await directly on the generator object itself, using symmetric tranfser.

Reply:
> Maybe I should pause the internals review and look at the interface. Why an iterator range interface? 

to maintain consistency with the interface proposed in https://wg21.link/P2502R2, i opted for a design that closely aligns with the standard. the proposal introduces `std::generator`, a standard library type implementing a coroutine generator that forms to the `std::ranges::input_range` model. by adhering to this proposed interface, we ensure that the developers familiar with the standard will feel at home when using the asynchronous generator. hopefully, this approach makes the asynchronous generator more intuitive and easier to use.

> It can't be be used in a range for loop. 

it can't due to the inherent difference between async generator and synchronous generator. and unfortunately, [Remove for `co_await` statement](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0664r8.html#35) was accepted. so we cannot use a ranged-for loop with async generator. but as explained above. probably we can preserve this interface? even though the async variant cannot allow the caller to use a nice "for" loop for enumerating the elements produced by the generator.

> How do you determine the generator is exhausted? Here you know it generates 2*n, but in general, you don't.

we do have tests exercising the "you don't" cases, for example, see `test_generator_drained_with_suspend`. i am quoting part of it as a reference

```c++
seastar::future<>
verify_fib_drained(coroutine::experimental::generator<int> actual_fibs, unsigned count) {
    auto expected_fibs = sync_fibonacci_sequence(count);
    auto expected_fib = std::begin(expected_fibs);

    auto actual_fib = co_await actual_fibs.begin();

    for (; actual_fib != actual_fibs.end(); co_await ++actual_fib) {
        BOOST_REQUIRE(expected_fib != std::end(expected_fibs));
        BOOST_REQUIRE_EQUAL(*actual_fib, *expected_fib);
        ++expected_fib;
    }
    BOOST_REQUIRE(actual_fib == actual_fibs.end());
}
```

> 
> I think we should go back to std::optional<T>

it is a step *back*.  the gist of the rewrite is to reuse the reference design proposed by https://wg21.link/P2502R2. so that:

* we don't need to track the emptiness or the free slots in the internal buffer. this is probably most important.
* we can save the overhead of calling move constructor or conversion.
* we can expose a similar interface to the one proposed by P2502R2

returning `std::optinoal<T>` defeats the purpose of the efforts of rewriting. especially it does not work with the first item in the list above.

Reply:
But what does `co_await ++it` mean? Shouldn't it be `co_await ++*it`? or `*co_await ++it`? 

Using iterators doesn't help, because there's no async for loop, and all the range algorithms don't work on such iterators.

How can you compare the iterator to the end iterator, before you co_await? How is the empty sequence represented?

Reply:
> But what does `co_await ++it` mean? 

move to (wait for) the next element.

> Shouldn't it be `co_await ++*it`? or `*co_await ++it`?

i think `*co_await ++it` works.

> 
> Using iterators doesn't help, because there's no async for loop, and all the range algorithms don't work on such iterators.

hmm, iterator helps with for loop. but i think it's also a way to represent a range. imaging we are walking from home to a bus stop, the start point is `begin(journey)`, and the bus stop is represented by `end(journey)`.

> 
> How can you compare the iterator to the end iterator, before you co_await? How is the empty sequence represented?

please see the code above, i am replicating it here:

```c++
for (; actual_fib != actual_fibs.end(); co_await ++actual_fib) {
   // ..
}
```

if we hit the sentinel in the first iteration, then it is an empty sequence.

Reply:
Let's say you're reading data from a file. Before you read anything, you don't know if the sequence ends.

However, the generator is returned immediately, before it has the chance to read anything. So we can't compare the begin/end iterators.

With yielding optionals, we return information on both the end-of-stream and data after awaiting. It's less nice than ranges, but it satisfies a requirement that ranges don't.

Reply:
@avikivity so the Fibonacci sequence is not representive enough. let me use a generator which reads a file then:
```c++
coroutine::experimental::generator<temporary_buffer<char>> read_file(std::string_view fn, size_t n) {
  auto file = open_file_dma(fn, open_flags::ro);
  auto in = make_file_input_stream(file);
  for (;;) {
    auto buf = co_await in.read_exactly(n);
    if (buf.size() < n) {
      co_return std::move(buf);
    } else {
      co_yield std::move(buf);
    }
  }
}

future<> cat(std::string_view fn) {
  auto i = co_await read_file.begin();
  for (; i != read_file.end(); co_await ++i) {
    auto& buf = *i;
    fmt::print("{}", std::string_view(buf.begin(), buf.end());
  }
}
```

what are you missing?

Reply:
It seems to work. But it's just odd. The begin/end methods don't return iterators. You don't compare begin() to end(), you compare co_await begin() to end(). It looks like a range but it isn't a range.

The interface is also odd. You await begin() and ++i, but not anything else. You can't do regular iterator stuff like *++i.

On the other hand, there's an advantage to being able to pass a reference, which isn't possible with std::optional\<T>. (there will be optional\<T&> in C++26).

Reply:
I see it's inspired by https://github.com/lewissbaker/cppcoro#async_generatort. Does it have users? We can learn from its experience.

Reply:
> It seems to work. But it's just odd. The begin/end methods don't return iterators. You don't compare begin() to end(), you compare co_await begin() to end(). It looks like a range but it isn't a range.
> 
> The interface is also odd. You await begin() and ++i, but not anything else. You can't do regular iterator stuff like *++i.
> 

yes, it's a known limit. i put this in the comment in the very beginning of the header file. see https://github.com/scylladb/seastar/pull/2218/commits/5ef1a6e6b1a777912bdfbdfdb96ea76a250f5651#diff-b7dc64e5af0d4dc90acdd0c4f909a940dbb2c74e88d65e503508ec1eb54253c5R48 . you are practically repeating it =( . i am quoting the comment here
```c++
// * Asynchronous Operations:
//   - generator::begin() is a coroutine, unlike P2502R2's synchronous approach
//   - generator::iterator::operator++() is a coroutine
//   - generator::iterator::operator++(int) is a coroutine
//   Note: Due to its asynchronous nature, this generator cannot be used in
//   range-based for loops.
// * Ranges Integration:
//   seastar's generator is not a std::ranges::view_interface. So it lacks
//   integration with the C++20 Ranges library due to its asynchronous operations.
```
> On the other hand, there's an advantage to being able to pass a reference, which isn't possible with std::optional<T>. (there will be optional<T&> in C++26).

you mean returning a reference from the producer or returning a reference to the consumer?

Reply:
> I see it's inspired by https://github.com/lewissbaker/cppcoro#async_generatort. Does it have users? We can learn from its experience.

yes, this implementation is also inspired by the cppcoro's `async_generator<T>` and folly generator implementation. i searched around with the keyword of `<cppcoro/async_generator.hpp>` . but couldn't find any users of it. i think, C++20 coroutines are a relatively new feature in the C++ world. generators, which are based on coroutines, are still gaining traction among developers. as a result, the number of programmers actively using generators remains limited. asynchronous generators, being an even more advanced concept, are even less commonly utilized in practical applications.

Reply:
@avikivity so are you okay with using an async iterator like: `for (auto it = co_await gen.begin(); it != gen.end(); co_await ++it) { auto val = *it; }`?

Reply:
Qt seems to use the same interface as proposed here for its `AsyncGenerator<T>` type (which is a good thing IMHO):

https://qcoro.dev/reference/coro/asyncgenerator/